### ISSUE TITLE
Corruption of files used during init(rd) prevents boot

### ISSUE THREAD

**Issue by unode** [NixOS/nixpkgs#61755]
## Issue description

**TLDR**: Files in `/var/lib/nixos`, `/etc/passwd` and `/etc/group` were partially filled with null-bytes after filling the root disk causing boot to fail/loop due to errors during init(rd).

----

I'm running NixOS with only one partition formatted in btrfs on a laptop.
At some point the disk became full and the laptop ran out of battery. I'm not sure if systemd managed to gently suspend/poweroff the system. The system was mostly idle at the moment of the crash.
The next boot no longer succeeded and picking an older boot configuration in grub didn't help either.

Whenever a boot was attempted I could see some init error messages flashing through the screen. They mentioned that files couldn't be chown'ed to root.root or other users.
The boot process continued regardless but some services failed (docker, dbus, ...). In particular, dbus failed to start but systemd kept retrying every minute or so in a never ending loop. 

At this point I started exploring recovery options.
Entering `single` mode through GRUB didn't help. I was presented with an error that the password store couldn't be read and I couldn't get a shell to the system. After some time of systemd reboots the system. 

Eventually after help from @cleverca22 on IRC I managed to get access to `/bin/sh` and start poking system logs. I realized the disk was full and promptly released some space. This was not sufficient to boot the system. Back to `/bin/sh` and after some hours of digging and running sanity checks on the filesystem 2 issues were found affecting system files.

The first, all files in `/var/lib/nixos/` were filled with null-bytes. This caused the perl scripts in initrd (`update-users-groups.pl`) to fail with JSON parsing errors. However these were not fatal and booting resumed. This was the cause of the `chown root.root` errors mentioned above. I have no explanation to how the files ended up filled with null bytes.

Additionally, both `/etc/passwd` and `/etc/groups` also had leading null bytes but still contained some regular content at the end of the file. It looked like the files were appended to regardless of their initial content.

Deleting all files in `/var/lib/nixos` and both files in `/etc/` after releasing some disk space was sufficient to repair and boot the system. After boot I had to correct user UID/GID which where not hardcoded in `configuration.nix` but otherwise all was back to normal. 

### Steps to reproduce

As mentioned above it's not entirely clear what caused the null-bytes but running out of space in the filesystem certainly played a role.

## Technical details

```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.1.0, NixOS, 19.03.git.19a5fbf (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2.2`
 - channels(unode): `"nixos-unstable-18.03pre117690.ab2cc75f78"`
 - nixpkgs: `/var/nixpkgs-channels/nixos`
```

### What could be improved

Some effort could be made in `update-users-groups.pl` or elsewhere during boot to validate the files in `/var/lib/nixos` and delete them if corrupted. Additionally `/etc/passwd` and `/etc/group` could be sanity checked.

It would also be useful to have a failsafe option in grub that at the very least provides a minimal functional shell for troubleshooting.

**Comment by edolstra** [comment]
Related to https://github.com/NixOS/nixpkgs/issues/26788.

**Comment by unode** [comment]
It's possible this has been fixed by #98544. Closing.

**Comment by iblech** [comment]
Sorry to reawaken this: Today I ran into this issue, on a family member's machine. It's nixpkgs unstable from a couple weeks ago. So #98544 did not fix this. That machine is on btrfs, don't know whether that's relevant.

Until somebody has tuits for investigating this further, I suggest we just continue recording cases here.

**Comment by iblech** [comment]
I appreciate your work on clarifying this! I observed:

- the null bytes in /etc/passwd
- reported size 0 of the files in /var/lib/nixos
- failure to boot
- I consider it likely that the btrfs partition was temporarily full, but when I was called to diagnose the problem several hundred MiB were free (and I did not encounter errors typical of a full btrfs system)
